plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py

```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env python
+
+import importlib
+import os
+import sys
+import urlparse
+import uuid
+import xapi.storage.api.v5.volume
+
+from xapi.storage import log
+from xapi.storage.libs import util
+from xapi.storage.libs.libcow.callbacks import VolumeContext
+from xapi.storage.libs.libcow.imageformat import ImageFormat
+from xapi.storage.libs.libcow.lock import PollLock
+from xapi.storage.libs.libcow.volume_implementation import Implementation as \
+    DefaultImplementation
+
+import zfsutils
+
+@util.decorate_all_routines(util.log_exceptions_in_function)
+class Implementation(DefaultImplementation):
+    "Volume driver to provide raw volumes from zvol's"
+
+    def create(self, dbg, sr, name, description, size, sharable):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        pool_name = meta["zpool"]
+
+        with VolumeContext(self.callbacks, sr, 'w') as opq:
+            image_type = ImageFormat.IMAGE_RAW
+            image_format = ImageFormat.get_format(image_type)
+            vdi_uuid = str(uuid.uuid4())
+
+            with PollLock(opq, 'gl', self.callbacks, 0.5):
+                with self.callbacks.db_context(opq) as db:
+                    volume = db.insert_new_volume(size, image_type)
+                    db.insert_vdi(
+                        name, description, vdi_uuid, volume.id, sharable)
+                    path = zfsutils.zvol_path(pool_name, volume.id)
+                    zfsutils.vol_create(dbg, path, size)
+
+                    vol_name = zfsutils.zvol_path(pool_name, volume.id)
+                    volume.vsize = zfsutils.vol_get_size(dbg, vol_name)
+                    if volume.vsize != size:
+                        log.debug("%s: VDI.create adjusted requested size %s to %s",
+                                  dbg, size, volume.vsize)
+                    db.update_volume_vsize(volume.id, volume.vsize)
+
+            vdi_uri = self.callbacks.getVolumeUriPrefix(opq) + vdi_uuid
+
+        return {
+            'key': vdi_uuid,    # FIXME check this
```

Comment:
the only requirement is that it should be unique? At least the common volume.py is doing the same. But I don't know if it is ok.

Reply:
yes, does not look useful to dig deeper at this stage